File: java/system_design/AdapterPattern.java
interface Payment { boolean pay(double a); }
class LegacySystem { boolean process(int cents) { return true; } }
/**
 * [Brief description]
 *
 * @param [param] [description]
 * @return [description]
 * @time O([complexity])
 */
class Adapter implements Payment { LegacySystem sys; Adapter(LegacySystem s) { sys = s; } public boolean pay(double a) { return sys.process((int)(a*100)); } }
/**
 * AdapterPattern - [Brief description]
 *
 * <p>OVERVIEW:
 * [Detailed explanation of what this class does]
 *
 * <p>COMPLEXITY:
 * <ul>
 *   <li>Time: [See method documentation]</li>
 *   <li>Space: O(n) where n is [the element count]</li>
 * </ul>
 *
 * <p>USAGE:
 * [How to use this class, with example]
 *
 * @author Interview Preparation
 * @since 1.0
 */

/**
 * [Brief description]
 *
 * @param [param] [description]
 * @return [description]
 * @time O([complexity])
 */
public class AdapterPattern { public static void main(String[] a) { Payment p = new Adapter(new LegacySystem()); System.out.println("Pay: " + p.pay(29.99)); } }

File: python/system_design/adapter_pattern.py
"""Adapter Pattern - Interface compatibility layer"""

from abc import ABC, abstractmethod


class Payment(ABC):
    """Target payment interface"""

    @abstractmethod
    def pay(self, amount: float) -> bool:
        raise NotImplementedError


class LegacyPaymentSystem:
    """Legacy payment system with different interface"""

    def process_payment(self, amount_cents: int) -> bool:
        """Legacy system takes cents, returns success/fail"""
        print(f"Legacy system processing ${amount_cents / 100:.2f}")
        return True


class PaymentAdapter(Payment):
    """Adapter to make legacy system compatible"""

    def __init__(self, legacy_system: LegacyPaymentSystem):
        self.legacy_system = legacy_system

    def pay(self, amount: float) -> bool:
        """Translate new interface to legacy interface"""
        amount_cents = int(amount * 100)
        return self.legacy_system.process_payment(amount_cents)


class ModernPaymentGateway(Payment):
    """Modern payment system"""

    def pay(self, amount: float) -> bool:
        print(f"Modern gateway processing ${amount:.2f}")
        return True


class ShoppingCart:
    """Shopping cart using payment interface"""

    def __init__(self, payment: Payment):
        self.payment = payment
        self.total = 0.0

    def add_item(self, price: float):

    """

    [Brief description of what this function does]


    Args:

        [param]: description


    Returns:

        [description of return value]


    Time: O([complexity])

    Space: O([complexity])

    """
        self.total += price

    def checkout(self) -> bool:
        return self.payment.pay(self.total)


if __name__ == "__main__":
    cart1 = ShoppingCart(ModernPaymentGateway())
    cart1.add_item(10.50)
    cart1.add_item(25.75)
    print("=== Modern Gateway ===")
    cart1.checkout()

    print()

    legacy = LegacyPaymentSystem()
    adapter = PaymentAdapter(legacy)
    cart2 = ShoppingCart(adapter)
    cart2.add_item(15.99)
    cart2.add_item(8.50)
    print("=== Legacy System (via Adapter) ===")
    cart2.checkout()
